### PR TITLE
Fixed asset.ready() for prefiltered cubemaps

### PR DESCRIPTION
**Pull request by yaustar**
Fixes https://github.com/playcanvas/engine/issues/4246

I confirm I have read the [contributing guidelines](https://github.com/playcanvas/engine/blob/master/.github/CONTRIBUTING.md) and signed the [Contributor License Agreement](https://docs.google.com/a/playcanvas.com/forms/d/1Ih69zQfJG-QDLIEpHr6CsaAs6fPORNOVnMv5nuo0cjk/viewform).

### CODE REVIEW

File: src/asset/asset.js [hunk 1]
```diff
@@ -464,7 +464,12 @@ class Asset extends EventHandler {
     ready(callback, scope) {
         scope = scope || this;
 
-        if (this.resource) {
+        // If cubemap asset doesn't need to load the faces, then resource is null as it references
+        // resources[0] that stores the faces. We need to check if the prefiltered assets are
+        // loaded instead
+        const assetLoaded = this.resource || (this.type === "cubemap" && !this.loadFaces && this.resources.length > 0);
```
Comment (yaustar) [inline src/asset/asset.js:470]: This feels very specific but not really sure how else to check

File: src/asset/asset.js [hunk 2]
```diff
@@ -464,7 +464,12 @@ class Asset extends EventHandler {
...
+
+        if (assetLoaded) {
```
Comment (slimbuck) [inline src/asset/asset.js:472]: Looks like this code was written years ago, perhaps before `loaded` flag was introduced. 

The correct code should be:
```suggestion
        if (this.loaded) {
```
This change would impact all assets though and will need thorough testing, but I think it's worth it.

Comment (yaustar) [reply src/asset/asset.js:472]: I will have another look at this as I vaguely remember there being an issue where something could be 'loaded' but the resource not be ready but can't remember the details

**Comment by Maksims** [comment]
Prefiltered cubemap has a complex "loaded" state:
1. Prefiltered data can be loaded only.
2. Faces can be loaded only.
3. Both can be loaded.

And depending on the situation, they all are valid cases. This is made to prevent unnecessary face loading (when they are not used) while loading prefiltered data for PBR.

**Comment by yaustar** [comment]
Had a test against the test cases that maksims listed above. In all cases, `.loaded` is true if the asset is used in the scene as skybox.

I also tested it as environment with a material and had the same results.

The case where this would potentially fail is if `loadFaces` is called on the cubemap asset after only the prefilter'ed data is loaded.

While the faces are loading, the cubemap asset loaded flag would still be true. It's an edge case that I think is fine to leave for now?

**Comment by slimbuck** [comment]
> Prefiltered cubemap has a complex "loaded" state:
> 
> 1. Prefiltered data can be loaded only.
> 2. Faces can be loaded only.
> 3. Both can be loaded.
> 
> And depending on the situation, they all are valid cases. This is made to prevent unnecessary face loading (when they are not used) while loading prefiltered data for PBR.

The complexity originates from the fact that a cubemap asset contains both a cubemap and/or prefiltered lighting data. These two things should not be the same asset. We can't easily change it now, but my hope is we can add a new prefiltered asset in future (perhaps storing the 2d equirect image directly, perhaps as png) and then deprecate the magical invisible prefiltered data on the cubemap.

**Comment by yaustar** [comment]
> Yeah tbh I kinda expected the cubemap asset to handle loaded correctly (since I rewrote it! 😀), but am less sure about all the other asset types. They'd need proper testing.

To be defensive, maybe we should do the following?

```
if (this.loaded || this.resource) {
```

File: src/asset/asset.js [hunk 3]
```diff
@@ -464,7 +464,7 @@ class Asset extends EventHandler {
     ready(callback, scope) {
         scope = scope || this;
 
-        if (this.resource) {
+        if (this.loaded) {
```
Comment (Maksims) [inline src/asset/asset.js:467]: Please, ensure logic described above for cubemaps are still present to prevent unnecessary downloads.

Comment (yaustar) [reply src/asset/asset.js:467]: I've double tested the following:

- Not preloaded, no prefiltered cubemap, on a material environment map that is used in the scene - Only faces are loaded
- Preloaded, no prefiltered cubemap, on a material environment map that is used in the scene - Only faces are loaded
- Not preloaded, prefiltered cubemap generated, on a material environment map that is used in the scene - Only prefiltered data is loaded
- Preloaded, prefiltered cubemap generated, on a material environment map that is used in the scene - Only prefiltered data is loaded

- Not preloaded, no prefiltered cubemap, asset.load called by code - Only faces are loaded
- Preloaded, no prefiltered cubemap, asset.load called by code - Only faces are loaded
- Not preloaded, prefiltered cubemap generated, asset.load called by code - Only prefiltered data is loaded
- Preloaded, prefiltered cubemap generated, asset.load called by code - Only prefiltered data is loaded

- Not preloaded, no prefiltered cubemap, used as skybox, mip level 2 - Only faces are loaded
- Preloaded, no prefiltered cubemap, used as skybox, mip level 2 - Only faces are loaded
- Not preloaded, prefiltered cubemap generated, used as skybox, mip level 2 - Only prefiltered data is loaded
- Preloaded, prefiltered cubemap generated, used as skybox, mip level 2 - Only prefiltered data is loaded
- Not preloaded, prefiltered cubemap generated, used as skybox, mip level 1 - Only faces and prefiltered data are loaded
- Preloaded, prefiltered cubemap generated, used as skybox, mip level 1 - Only faces and prefiltered data are loaded

Have I missed any other possible test cases?

Comment (yaustar) [reply src/asset/asset.js:467]: I can't see any where in the code base where .ready() is used for cubemap texture loading internally in the engine so we should have the same behaviour as before

**Comment by yaustar** [comment]
Created a project that has all the assets and tests ready() against them. 

https://playcanvas.com/project/927453/overview

**Review by slimbuck — APPROVED** [review]

**Pull request merged**